File: python/xenarch/merchant/_idempotency.py
"""Idempotency-Key minting for link creation.

The platform honors ``Idempotency-Key`` on ``POST /v1/links`` and returns the
cached link only when the *same body + key* arrives twice — so a fresh key per
create dedupes a network-level retry of one create request. (A re-run of
``create(params)`` mints a fresh nonce, i.e. a different body, so it makes a
new link; cross-invocation dedup isn't possible without a stable nonce.) The
jsonl is an append-only audit trail of issued keys. Mirrors the CLI's
``idempotency.ts``.
"""
from __future__ import annotations

import json
import os
import uuid
from pathlib import Path

_CONFIG_DIR = Path(os.path.expanduser("~")) / ".xenarch"


def new_key() -> str:
    """A fresh idempotency key (UUID4 hex, ``idem_`` prefixed)."""
    return f"idem_{uuid.uuid4().hex}"


def record(key: str, *, created_at: int, config_dir: str | Path | None = None) -> None:
    """Append a key to the local audit log. Best-effort — never raises."""
    directory = Path(config_dir) if config_dir else _CONFIG_DIR
    try:
        directory.mkdir(parents=True, exist_ok=True, mode=0o700)
        line = json.dumps({"key": key, "created_at": created_at})
        with (directory / "idempotency.jsonl").open("a", encoding="utf-8") as fh:
            fh.write(line + "\n")
    except OSError:
        # An unwritable home dir must not block a create.
        pass

File: js/sdk/src/idempotency.ts
/**
 * Idempotency-Key minting for link creation.
 *
 * The platform honors `Idempotency-Key` on `POST /v1/links` and returns the
 * cached link only when the *same body + key* arrives twice — so a fresh key
 * per create dedupes a network-level retry of one create request. (Re-running
 * `create` mints a fresh nonce, i.e. a different body, so it makes a new link;
 * cross-invocation dedup isn't possible without a stable nonce.) The audit log
 * is best-effort and Node-only — skipped silently on edge/browser runtimes.
 */
export function newIdempotencyKey(): string {
  return `idem_${crypto.randomUUID().replace(/-/g, "")}`;
}

/** Append a key to `~/.xenarch/idempotency.jsonl`. Best-effort; never throws. */
export async function recordIdempotency(entry: {
  key: string;
  link_id?: string;
}): Promise<void> {
  try {
    // Dynamic imports so non-Node runtimes (edge/browser) don't fail to load.
    const os = await import("node:os");
    const path = await import("node:path");
    const fs = await import("node:fs/promises");
    const dir = path.join(os.homedir(), ".xenarch");
    await fs.mkdir(dir, { recursive: true, mode: 0o700 });
    const line =
      JSON.stringify({ ts: new Date().toISOString(), ...entry }) + "\n";
    await fs.appendFile(path.join(dir, "idempotency.jsonl"), line, {
      mode: 0o600,
    });
  } catch {
    // No home dir / not Node / unwritable — auditing must not block a create.
  }
}
